[FIX] account: handle non-alphanumeric journal codes in European payment references

path-odoo opened the pull request:
Currently, an error occurs when confirming a customer invoice that uses a journal with an invalid configuration.

**Steps to Reproduce:**
1) Install Accounting app.(with Demo)
2) Navigate to Accounting>Configuration>Journals
3) Open existing 'Sales' journal and make following changes:
- set Sequence Prefix: INV-
- In the Advanced Settings Page, set 'Communication Standard' as 'European' and
   save the journal.
4) Create a Customer Invoice with this Sales Journal and click on `Confirm`.

**Error :**
`ValueError: invalid literal for int() with base 36: '-'`

**Root Cause:**
Since [this commit](https://github.com/odoo/odoo/pull/212169/commits/e4a09467a20f282d28baa09091c10a6aea6d0094#diff-cc13d9842e166c12738b659e0478ceb8b1b15734442c1e090252919e7efb6ed7), On following above steps, The value of 'number' received at [1]  looks like 'INV-000003' due to which on further computation of this value at [1] causing an error.

**Fix:**
prevent crash by adding a additional check on methods
`_get_invoice_reference_euro_invoice` and `_get_invoice_reference_euro_partner`
to validate the journal's short code.

[1]- https://github.com/odoo/odoo/blob/2646790e7e31b4c600e39f2d5c87f0744b1578b0/addons/account/tools/structured_reference.py#L20-L26

**sentry-6860586611**

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/225987.png)](https://mergebot.odoo.com/odoo/odoo/pull/225987)

path-odoo commented:
Hello @odoo/sentry_reviewers,
Could I get a review here?
Thanks :smile:!

aboo-odoo reviewed (commented):
Hey! I think a constrain is a bit overkill, it truly limits what users can do. I believe this can be solved simply by using
```
self.journal_id.code**.encode()**.isascii()
```
in `_get_invoice_reference_euro_{invoice,partner}(...)`

For example: with a journal code like INVα (the alpha greek character). The code is stored as INVa for which `isascii` returns true. But with the `encode`, it becomes `INV\xce\xb1` with is not ascii.

path-odoo commented:
> Hey! I think a constrain is a bit overkill, it truly limits what users can do. I believe this can be solved simply by using
> 
> ```
> self.journal_id.code**.encode()**.isascii()
> ```
> 
> in `_get_invoice_reference_euro_{invoice,partner}(...)`
> 
> For example: with a journal code like INVα (the alpha greek character). The code is stored as INVa for which `isascii` returns true. But with the `encode`, it becomes `INV\xce\xb1` with is not ascii.

Hello @aboo-odoo :wave:,
Thanks for your suggestion! I tried implementing it, but the system was still crashing because the `-`(hypen) was treated as an ASCII character. This allowed it to pass the condition and return `True`.

So, I removed my previous fix that used a constraint and instead added the following changes:

Before:
```PYTHON
journal_identifier = self.journal_id.code if self.journal_id.code.isascii() else self.journal_id.id
```
After:
```PYTHON
journal_identifier = self.journal_id.code if self.journal_id.code.isascii() and self.journal_id.code.isalnum() else self.journal_id.id
```

I added an extra check with **'self.journal_id.code.isalnum()'**, This way, we can allow users to create custom short codes without restricting them and also prevent the crash.

I’ve updated the test case accordingly. Let me know your thoughts on this.

Thanks :smile:!

aboo-odoo commented on addons/account/tests/test_account_journal.py:
```diff
@@ -347,8 +379,9 @@ def test_non_latin_journal_code_payment_reference(self):
+        expected_id = str(invoice_non_latin.journal_id.id)
         ref_parts_non_latin = invoice_non_latin.payment_reference.split()
-        self.assertEqual(ref_parts_non_latin[1][:3], str(invoice_non_latin.journal_id.id), "The reference should start with " + str(invoice_non_latin.journal_id.id))
+        self.assertEqual(ref_parts_non_latin[1][:len(expected_id)], str(invoice_non_latin.journal_id.id), "The reference should start with " + str(invoice_non_latin.journal_id.id))
```
```suggestion
        self.assertEqual(ref_parts_non_latin[1][:len(expected_id)], expected_id, "The reference should start with " + expected_id)
```

aboo-odoo reviewed (commented):
Nice thank you! Small comment then we'll be good :+1:

aboo-odoo reviewed (approved):
Ship it :+1: 

@robodoo r+
